decentraland/explorer
# fix: error reporting overhaul

# What? 

This PR implements most of the requirements of the [error reporting RFC](https://docs.google.com/document/d/1eCHxBpvnFc-fvEXf540y7unUr0mXhQ6TmidkkUnTn2E/edit#heading=h.o921fa8bd20p)

In a nutshell:
- An additional `context` field is added for easy filtering
- Manual throws (saga errors, unity's error handler, etc) are uploaded to Rollbar as critical errors.
- Unity errors include a BIG payload with detailed client state as JSON (this weights around 25KB or more). For more details on the payload contents check the RFC.
- Refactored all our throw instances to have some form of `context` data.
- Hardware capability errors (like no webgl, etc) aren't reported to rollbar, but they are tracked on analytics.
- `comms_error_x` analytics event changed to `error_comms_retry` and the number of retries is included in the error payload.

# How to test

Go to console and jot the following:
``` 
clientDebug.TestErrorReport(`your critical error of choice`, `renderer#errorHandler`)
```

Now, you'll see the error appear in Rollbar dashboard.

To visualize the full unity payload, you have to use the "Explore as RQL query" on the ocurrences tab:

![image](https://user-images.githubusercontent.com/6875814/120706001-ea493b00-c48e-11eb-9f19-cb283f1d558b.png)

Then change the visualization mode and voila:

![image](https://user-images.githubusercontent.com/6875814/120706080-0a78fa00-c48f-11eb-9c46-6d3d95113f6f.png)

Technical details:

- now, `ReportFatalError` will only report the error to Rollbar and accept Error objects as parameters, so rollbar has the stack and other useful data
- now, `BringDownClientAndShowError` is exported and is in charge of send the tracking event for the popup, as the tracking data can be useful. `ReportFatalError` doesn't send tracking events anymore. 
- For compatibility errors, we just call `BringDownClientAndShowError`. We don't report compatibility errors to Rollbar.
- `ReportFatalError` now report the errors as critical.
- introduced the `ErrorContext` namespace so the errors can be classified in Rollbar's dashboard.
- add placeholder to inject renderer's and comms' payloads.
- removed the reporting for renderer messages queue as this is not useful at all
- modified `generic_error` payload to match Rollbar payload
